packages/adminrouter/extra/src/README.md

```diff
@@ -264,28 +264,178 @@ it, but it does not seem to be justified. For example:
+Unfortunately, Mesos as of now does not enforce framework name uniqueness and
+thus it is possible to launch two frameworks with the same name. AR in such
+case will route in a deterministic manner. Root Marathon by default enforces
+unique names for all the tasks and frameworks it has under control.
+
+Please check the [Limitations](#limitations) section for more details.
+
+### Operation
+Under the hood, the `/service` endpoint follows non-trivial algorithm in order to
+resolve request path into IP + Port pair.
+
+#### Recursive resolving
+There is no way to tell which part of the path is namespace+service name and
+which one is just a resource address. For example, assumming that the request
+path is `/foo/bar/baz/picture.jpg` we can have:
+* namespace and name: `foo`, resource address: `/bar/baz/picture.jpg`
+* namespace and name: `foo/bar`, resource address: `/baz/picture.jpg`
+* namespace and name: `foo/bar/baz`, resource address: `/picture.jpg`
+* namespace and name: `foo/bar/baz/picture.jpg`, resource address: `/`
+
+So the algorithm iteratively tries to resolve each "candidate" namespace and name
+and assume that first-match wins.
+
+Assuming that we deployed `foo/bar` NGINX task, we will see two iterations:
+* first one with `foo` candidate namespace+service, which is not present in the
+cluster
+* second one with `foo/bar` candidate namespace+service name, which results in
+request being routed to correct backend. In case when user is trying to access
+an non-existent service, there are going to be four iterations, each one
+unsuccessful. After the fourth one the code will give up.
+
+There is of course a limit on the number of iterations that AR makes. Currently,
+it is set to `10`. This means that a service cannot nest more than nine times and
+that everything after the tenth segment is always treated as a resource component.
+
+#### Resolving individual component
+During each iteration described in previous paragraph, AR tries to resolve the
+component using the data stored in the cache and if necessary - from MesosDNS:
+* first step is to check if there is a root Marathon task that has lebels set
+  and these lebels:
+  * `DCOS_SERVICE_NAME` label contents is equal to given `namespace+service` string
+  * `DCOS_SERVICE_SCHEME` label is set
+  * `DCOS_SERVICE_PORT_INDEX` label is set
+  * task is in state `TASK_RUNNING`
+  If all the conditions above are met, request is routed using root Marathon
+  data. Please check the [section below](#recommended-way-to-expose-services)
+  for details on the meaning of these labels.
+* if the name of the service couldn't be resolved using root Marathon data, AR
+  scans the list of Mesos frameworks in search for the one with either framework
+  ID (higher priority) or framework name equal(lower priority) to
+  `namespace+service` string. If found, depending on the value of `webui_url` field:
+  * field contains an empty string - algorithm tries to resolve the
```

Comment:
```
  347 * if the name of the service couldn't be resolved using root Marathon data, AR
~ 348   scans the list of Mesos frameworks in search for the one with either
~ 349   framework ID (higher priority) or framework name equal(lower priority) to
~ 350   application ID string. If found, the next steps depend on the value of
~ 351   `webui_url` field:
~ 352   * field contains an empty string - algorithm tries to resolve the application
~ 353     ID using MesosDNS
~ 354   * field is `null` - resolving process for given `namespace and name` ends,
~ 355     MesosDNS step is NOT executed
+ 356   * field contains a valid address - request is proxied to the address pointed
+ 357     by the field contents
```
The idea here is to present a list of options:
* empty field - blah, blah...
* field is null  - blah, blah...
* field is an url  - blah, blah...